### PR TITLE
[controls] fix memory leak with `CarouselView` & `INotifyCollectionChanged`

### PR DESCRIPTION
**Pull request by jonathanpeppers**
Context: https://github.com/dotnet/maui/issues/17726

In investigating #17726, I found a memory leak with `CarouselView`:

1. Have a long-lived `INotifyCollectionChanged` like
   `ObservableCollection`, that lives the life of the application.

2. Set `CarouselView.ItemsSource` to the collection.

3. `CarouselView` will live forever, even if the page is popped, etc.

I further expanded upon `MemoryTests` to assert more things for each
control, and was able to reproduce this issue.

To fully solve this, I had to fix issues on each platform.

### Android ###

`ObservableItemsSource` subscribes to the `INotifyCollectionChanged`
event, making it live forever in this case.

To fix this, I used the same technique in 58a42e59:
`WeakNotifyCollectionChangedProxy`. `ObservableItemsSource` is used for
`ItemsView`-like controls, so this may fix other leaks as well.

### Windows ###

Windows has the same issue as Android, but for the following types:

* `CarouselViewHandler` subscribes to `INotifyCollectionChanged`

* `ObservableItemTemplateCollection` subscribes to
  `INotifyCollectionChanged`

These could be fixed with `WeakNotifyCollectionChangedProxy`.

Then I found another issue with `ItemTemplateContext`, it holds a strong
reference to `BindableObject Container`, the `CarouselView`. I found
this kept the `CarouselView` alive indefinitely, but changing to a
`WeakReference` solved this issue.

These three types are used for other `ItemsView`-like controls, so this
may fix other leaks as well.

### iOS/Catalyst ###

These platforms suffered from multiple circular references:

* `CarouselViewController` -> `CarouselView` -> `CarouselViewHandler` ->
  `CarouselViewController`

* `ItemsViewController` -> `CarouselView` -> `CarouselViewHandler` ->
  `ItemsViewController` (`CarouselViewController` subclasses this)

* `CarouselViewLayout` -> `CarouselView` -> `CarouselViewHandler` ->
  `CarouselViewLayout`

The analyzer added in 2e656264 did not yet catch these because I only
added the analyzer so far for `Microsoft.Maui.dll` and these issues were
in `Microsoft.Maui.Controls.dll`. In a future PR, we can proactively try
to add the analyzer to all projects.

### Conclusion ###

Unfortunately, this does not fully solve #17726, as there is at least
one further leak on Android from my testing. But this is a good step
forward.

### CODE REVIEW

File: src/Controls/src/Core/Handlers/Items/Android/ItemsSources/ItemsSourceFactory.cs [hunk 1]
```diff
@@ -20,7 +20,7 @@ public static IItemsViewSource Create(IEnumerable itemsSource, BindableObject co
 				case IList list when itemsSource is INotifyCollectionChanged:
 					return new ObservableItemsSource(new MarshalingObservableCollection(list), container, notifier);
 				case IEnumerable _ when itemsSource is INotifyCollectionChanged:
-					return new ObservableItemsSource(itemsSource as IEnumerable, container, notifier);
+					return new ObservableItemsSource(itemsSource, container, notifier);
```
Comment (jonathanpeppers) [inline src/Controls/src/Core/Handlers/Items/Android/ItemsSources/ItemsSourceFactory.cs:23]: This `as` was just extra, the `itemsSource` is already `IEnumerable`.

**Comment by jonathanpeppers** [comment]
The new test fails on iOS & Windows, looking into it.

**Comment by jonathanpeppers** [comment]
This one ended up being a big mess: different problems on each platform...

If this is green, I'll rebase and explain in a better PR description/commit message.

File: src/Controls/src/Core/Handlers/Items/iOS/ItemsViewController.cs [hunk 2]
```diff
@@ -15,35 +16,40 @@ public abstract class ItemsViewController<TItemsView> : UICollectionViewControll
...
 		Func<UICollectionViewCell> _getPrototype;
 		CGSize _previousContentSize = CGSize.Empty;
 
+		[UnconditionalSuppressMessage("Memory", "MA0002", Justification = "Proven safe in test: MemoryTests.HandlerDoesNotLeak")]
 		UIView _emptyUIView;
 		VisualElement _emptyViewFormsElement;
 		Dictionary<object, TemplatedCell> _measurementCells = new Dictionary<object, TemplatedCell>();
 		List<string> _cellReuseIds = new List<string>();
 
+		[UnconditionalSuppressMessage("Memory", "MA0002", Justification = "Proven safe in test: MemoryTests.HandlerDoesNotLeak")]
```
Comment (rmarinho) [inline src/Controls/src/Core/Handlers/Items/iOS/ItemsViewController.cs:47]: So maybe get the other one with rename of the error in first ?

Comment (jonathanpeppers) [reply src/Controls/src/Core/Handlers/Items/iOS/ItemsViewController.cs:47]: Let me just update this one to use `MEM` 👍

**Review by rmarinho — APPROVED** [review]

**Pull request merged**